tiledb/sm/rest/rest_client.cc

```diff
@@ -675,6 +690,17 @@ Status RestClient::get_query_est_result_sizes(const URI& uri, Query* query) {
       query, serialization_type_, true, returned_data);
 }
 
+std::string RestClient::redirect_uri(
+    const std::string& array_ns, const std::string& array_uri) {
+  const std::string redirect_key = array_ns + ":" + array_uri;
+  std::unique_lock<std::mutex> rd_lck(redirect_mtx_);
+  std::cerr << "Redirected key:" << redirect_key << std::endl;
+  std::unordered_map<std::string, std::string>::const_iterator cache_it =
+      redirect_meta_.find(redirect_key);
+  return (cache_it == redirect_meta_.end()) ? rest_server_ :
+                                              "https://" + cache_it->second;
```

Comment:
It is not safe to hard-code `https://` here (cc: @Shelnutt2 since I talked to him about this)

Instead, we need to parse the scheme (either http or https) during the direction. Here's an example of a response from the command line tool:
```
< HTTP/2 302
< date: Wed, 09 Dec 2020 13:10:36 GMT
< content-length: 0
< set-cookie: __cfduid=d40ce38b8790d7cb5c4d3477525feaf231607519436; expires=Fri, 08-Jan-21 13:10:36 GMT; path=/; domain=.tiledb.workers.dev; HttpOnly; SameSite=Lax
< location: https://eu-west-2.aws.api.tiledb.com/v1/arrays/TileDB-Inc/quickstart_dense-eu-west-2/metadata
< x-tiledb-full-region-domain: eu-west-2.aws.api.tiledb.com
< x-tiledb-provider: aws
< x-tiledb-region: eu-west-2
< x-tiledb-sub-domain: eu-west-2.aws
< cf-request-id: 06e93918480000096ff3b75000000001
< expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
< report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report?s=Xk3hzp3ZTuNw4NjGqQ5m%2FH2g4N8Jehc194rGjTtgEQu9JcQ5ksMDex2AaiyHfws41hkvfjW4c3r0m7qr81vjzwY2cdNaK%2F0GXbC%2BPqdNTPZ9SVkXT2CUoPLrfDPg3eJzq3BCcGwfu3xqtWGZIGWV5r%2F7yDfKAzm4T2%2FBZFw%3D"}],"group":"cf-nel","max_age":604800}
< nel: {"report_to":"cf-nel","max_age":604800}
< server: cloudflare
< cf-ray: 5feef7a079d3096f-IAD
<
```

We can get the scheme by parsing the "location: " header.

Seth pointed out that the curl library has a routine for parsing this already, see https://curl.se/libcurl/c/curl_url_get.html

---

I think it would be easier to parse `location` instead of `x-tiledb-full-region-domain`. We should be able to parse the scheme+uri, array namespace, and array name all in one pass on `location`. I know this will add a lot more time dev time, sorry!